### DISCUSSION TITLE
Language interoperability

### DISCUSSION THREAD

**Discussion by kesavkolla** [facebookincubator/velox#2508] [General]
Is there any plan to provide other programming language bindings for velox? I am looking for rust bindings so that I can use velox in rust.

**Comment by pedroerp** [comment]
@kgpai has started looking into Python bindings and there were discussions in the past with the Intel team about generalizing the Java bindings from Gluten, but we haven't discussed about Rust yet (though it would be nice to support). 

Is it something you would be interested in contributing?

**Comment by kesavkolla** [comment]
@pedroerp If there is some architectural guidence on how to go about providing language bindings then I can do rust bindings. I think currently the language bindings are happening outside the velox and I don't know how are they implementing. If there is some structure from velox team then other language bindings can flurish.

**Comment by kgpai** [comment]
@kesavkolla Once we have the python bindings out ; you can probably use that as a reference for rust . We are still working on the api's and trying to determine the public / private interface , so its hard to provide guidance on bindings just yet.

**Comment by kesavkolla** [comment]
@kgpai I understand that. In general we should be using something similar to Arrow C interface. This will give seemless interactions of data between programming language to native C.

**Comment by kesavkolla** [comment]
Is it possible to develop WASI layer for velox? That way we can integrate velox into any programming language. Today we have to work with C/C++ in native programming language integrations. If there is a WASI compatible layer then people can use in any programming language. I hope runtimes like wasmer or wasmtime can work with very low overhead.

**Comment by kesavkolla** [comment]
Is there any movement on this front? Is there any easy path that developers can use velox from other programming languages?

**Comment by kgpai** [comment]
Hi Kesav, Firstly, Apologies for the delayed response. Atm I am not aware of an plans to add WASI layer for Velox. You are welcome to contribute this by creating an issue and detailing the design etc. We have started with python bindings for Velox and have created PyVelox with rudimentary bindings (atm only expression eval and simple flat vectors) though we plan to add a lot more to this in the immediate short term. 
So to answer your question atm only Python is officially supported , also the support is relatively rudimentary, but we are working on it and will add lot more stuff (encodings, complex vectors, arrow/velox conversion, operators etc..) . 
Again we would welcome any contribution to better support this across other languages.

**Comment by philippemnoel** [comment]
Hi there! I’m interested in using Velox in Rust as well and will be looking on implementing bindings. Would love to work on them alongside the core Velox team and contribute them directly to Velox if that would be the wish of the team.

Is this something others would be willing to work on alongside myself and my team? Any tips other than inspire ourselves from PyVelox?

**Reply by kgpai** [reply]
Hi Phillip, 

 It will be great to have Rust bindings as well ! I dont have much experience with Rust but I can definitely help in any way I can to get you folks get started. Right now the only other language binding we have is python bindings with PyVelox , so feel free to leverage that in any way to help. 

A very rough plan would be : 
1. Join our slack channels / Attend the Open source meetings (They are advertised in Slack ) , so you can introduce yourselves to the community and know the folks to reach out to help.
2. Start with creating a rudimentary Rust package that binds to say the Velox type system.

Might I be forward and ask what the motivation for adding Rust support is ? I ask this because Velox is being rapidly updated and thus it would be good to understand what the long term support model for the Rust package would be. 

[ At this moment I am not sure if Rust should be part of our CI / build systems ,mainly because none of the core maintainers have Rust experience and we havent yet seen any customer ask for it, so as things rapidly evolve we would need someone to maintain the Rust bindings. In that light, I would encourage you to submodule Velox and have the rust package be built against that and as it gets more mature we could consider it making part of the core Velox repo. I will be interested in what other folks feel about this though ]


cc: @pedroerp

**Reply by philippemnoel** [reply]
Hi @kgpai! 

Thank you, that sounds great. We can get started and show you our work as we go. I think the submodule+separate repo approach makes sense for now, if we do use Velox we'd be happy to maintain the bindings, I suspect. I would need to confirm on that, though, and it would depend on how heavy a usage we do. Also note that we probably wouldn't make bindings for everything, just what we need, as a start. Maybe that's where others can jump in as needed for their use cases.

Re: Slack channel. I messaged velox@meta.com as per the README but never heard back. Would you be willing to invite me to the Slack, please? philippemnoel@gmail.com is my email. I'll be glad to attend the OS meetings.

We're considering bundling Velox into a Postgres extension built in Rust. This is where the interest is for bindings :)

Lmk!

**Reply by pedroerp** [reply]
Slack invite sent

**Reply by philippemnoel** [reply]
Thank you!

**Reply by ozgrakkurt** [reply]
Hey!! was there any progress on this? I would like to help also

**Reply by pedroerp** [reply]
I'm not aware of any progress on this direction, but would indeed be nice to provide Rust bindings.
